training/src/main/scala/com/airbnb/aerosolve/training/GradientUtils.scala

```diff
@@ -0,0 +1,112 @@
+                  dim :Int,
+                  lambda : Double) = {
+    var gradientNorm = 0.0
+    var featureCount = 0
+    gradients.foreach(kv => {
+      val (key, gradient) = kv
+      val featureMap = weightVector.get(key._1)
+      if (featureMap != null) {
+        val weight = featureMap.get(key._2)
+        if (weight != null) {
+          // Just a proxy measure for convergence.
+          gradientNorm = gradientNorm + gradient.grad.dot(gradient.grad)
+          val scale = 2.0 / math.max(1e-6, gradient.featureSquaredSum)
+          weight.multiplyAdd(-scale.toFloat, gradient.grad)
+          val hingeScale = 1.0 - lambda * scale / math.sqrt(weight.dot(weight))
+          if (hingeScale <= 0.0f) {
+            // Weights may get re-activated ... filter at the end.
+            weight.setZero(dim)
+          } else {
+            featureCount = featureCount + 1
+            weight.scale(hingeScale.toFloat)
+          }
+        }
+      }
+    })
+    log.info("Sum of Gradient L2 norms = " + gradientNorm)
+    log.info("Num active features = " + featureCount)
+  }
+
+  // Improved RPROP- algorithm
+  // https://en.wikipedia.org/wiki/Rprop
+  // http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.21.3428
+  def rprop(gradients : Map[(String, String), GradientContainer],
+            prevGradients : Map[(String, String), GradientContainer],
+            step : scala.collection.mutable.HashMap[(String, String), FloatVector],
+            weightVector : java.util.Map[String,java.util.Map[String,com.airbnb.aerosolve.core.util.FloatVector]],
+            dim :Int,
+            lambda : Double) = {
+    // Some fixed parameters for rprop, they don't matter much as the algorithm adapts them
+    // so just keep them fixed here.
+    // Step size parameters
+    val deltaInitial = 0.01f
+    val deltaMax = 50.0f
+    val deltaMin = 1e-6f
+    // Change in step size parameters
+    val etaPlus = 1.2f
+    val etaMinus = 0.5f
+
+    gradients.foreach(kv => {
+      val (key, gradient) = kv
+      val featureMap = weightVector.get(key._1)
+      if (featureMap != null) {
+        val weight = featureMap.get(key._2)
+        if (weight != null) {
+          val prev = prevGradients.get(key)
+          val prevGrad = if (prev.isEmpty) new FloatVector(dim) else prev.get.grad
+          val currOpt = step.get(key)
+          // Create a new step vector if we don't have one.
+          if (currOpt.isEmpty) {
+            val tmp = new FloatVector(dim)
+            tmp.setConstant(deltaInitial)
+            step.put(key, tmp)
+          }
+          val currStep = step.get(key).get
+          for (i <- 0 until dim) {
+            val prod = prevGrad.values(i) * gradient.grad.values(i)
+            if (prod > 0) {
+              currStep.values(i) = math.min(currStep.values(i) * etaPlus, deltaMax)
+            } else if (prod < 0) {
+              currStep.values(i) = math.max(currStep.values(i) * etaMinus, deltaMin)
```

Comment:
I am not very familiar with this algorithm, but according to the paper, when `prod < 0`, weight should not be updated: `Additionally, in case of a change in sign, there should be no adaptation in the succeeding learning step`